components/status_dropdown/status_dropdown.jsx

```diff
@@ -37,35 +41,101 @@ export default class StatusDropdown extends React.PureComponent {
         status: UserStatuses.OFFLINE,
     }
 
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            openUp: false,
+            width: 0,
+        };
+    }
+
+    getCurrentDateTime = (tz, enable) => {
+        let currentDate;
+        if (enable) {
+            if (tz.useAutomaticTimezone) {
+                currentDate = getCurrentDateTimeForTimezone(tz.automaticTimezone);
+            } else {
+                currentDate = getCurrentDateTimeForTimezone(tz.manualTimezone);
+            }
+        }
+        return currentDate;
+    }
+
+    dndTimes = ['30 mins', '1 hour', '2 hours', 'Today', 'Tomorrow', 'Custom']
+
     isUserOutOfOffice = () => {
         return this.props.status === UserStatuses.OUT_OF_OFFICE;
     }
 
-    setStatus = (status) => {
+    setStatus = (status, dndEndTime) => {
         this.props.actions.setStatus({
             user_id: this.props.userId,
             status,
+            dnd_end_time: dndEndTime,
         });
     }
 
     setOnline = (event) => {
         event.preventDefault();
-        this.setStatus(UserStatuses.ONLINE);
+        this.setStatus(UserStatuses.ONLINE, '');
     }
 
     setOffline = (event) => {
         event.preventDefault();
-        this.setStatus(UserStatuses.OFFLINE);
+        this.setStatus(UserStatuses.OFFLINE, '');
     }
 
     setAway = (event) => {
         event.preventDefault();
-        this.setStatus(UserStatuses.AWAY);
+        this.setStatus(UserStatuses.AWAY, '');
     }
 
-    setDnd = (event) => {
+    setDnd = (event, index) => {
         event.preventDefault();
-        this.setStatus(UserStatuses.DND);
+
+        var currentTime = this.getCurrentDateTime(this.props.userTimezone, this.props.enableTimezone).getTime();
+        var endTime = new Date().getTime();
+        switch (index) {
+        case 0:
+            endTime = new Date(currentTime + (1 * 30 * 60000));
+            break;
+        case 1:
+            endTime = new Date(currentTime + (2 * 30 * 60000));
+            break;
+        case 2:
+            // add 2 hours in current time
+            endTime = new Date(currentTime + (4 * 30 * 60000));
+            break;
+        case 3:
+            // set hours of date to point to last moment of the day
+            endTime.setHours(23, 59, 59, 999);
+            break;
+        case 4:
+            // set hours of date to point to last moment of the day
+            // and add 24 hours to it to point to last moment of tomorrow
+            currentTime.setHours(23, 59, 59, 999);
+            endTime = new Date(currentTime + (48 * 30 * 60000));
```

Comment:
using `48 * 30 * 60000` might be misleading here ... Please use `24 * 60 * 60 * 1000` (h * m * s * ms).
It is the more approach most devs use and know.